include/deal.II/base/mpi.h

```diff
@@ -235,9 +237,12 @@ namespace Utilities
      * array is the sum over the i-th entries of the input arrays from each
      * processor.
      *
+     * This function can also be used for ArrayView objects initialized with
+     * data living in CUDA memory if MemorySpace::CUDA is chosen.
+     *
      * Input and output arrays may be the same.
      */
-    template <typename T>
+    template <typename T, typename MemorySpaceType = dealii::MemorySpace::Host>
```

Comment:
I can't claim that I really understand what this is about, but how is this interpreted syntactically: that the object of type `T` is stored in device memory? That the function is called from the device? That the array view lives in the host memory space, but that the *pointer* that it stores points into device memory? (In which case the question is how one would access it.) Do you think you could clarify?

Reply:
We assume that we have an `ArrayView` on the CPU that stores a pointer to data either on the CPU or on the GPU. The function is called from the CPU. In case the data is stored on the GPU, we can't access it directly.

Reply:
I think I mentioned this in another context already, but I think that's an abuse of the `ArrayView` class. If you can't see the data, then it's not a "view" -- it's really something else. I would prefer if there were a different class for that then, and in that case you also don't have to touch the existing MPI interfaces.

Reply:
At least for `LinearAlgebra::distributed::Vector` the challenge is to write code that both works for data living on the CPU and the GPU. Using `ArrayView` for both allows for code that doesn't need to be specialized too much. Of course, I could also just provide overloads for all the methods of the `Partitioner` class for raw pointers and use these. On the other hand, I try to avoid raw pointer interfaces as much as possible. If I try to represent an array having a pointer pointing to the first element and its length at hand, `ArrayView` seems to be a natural choice.

If were are storing the storage location inside the `VectorView` object, we don't need to touch existing MPI interfaces either (just the implementation). Having a different class name makes using the `MemorySpaceType` template parameter really ugly.

Would you like a different class that is essentially equal to `VectorView` but without an `operator[]` member function better?

For me, it is not too much a stretch to say that the data in `ArrayView` is accessible or not depending on the storage location of the data represented. Of course, I am biased here. What do others think?

Reply:
ArrayView is `a class that represents a window of memory locations of type ElementType
and presents it as if it was an array that can be accessed via an operator[]` Which is exactly what we want. The only caveat is that the data lives on the device not on the host so you cannot access it from the host. That means that to use `operator[]` you need to be careful and store the address of the element in a pointer and then use this pointer in a kernel. So to access an element in the `ArrayView`, you need to do something like this:
```cuda
int *a = &my_array_view[2];
my_kernel<<<1,1>>>(a);
```
We could also just use a pair of a raw pointer and a size and use `using` to give it another name and hide the fact that it is a pair.

Reply:
I guess the question here depends on what an `ArrayView` is supposed to be in the context of the different memory spaces and whether, if I understand @bangerth correctly, it would be more appropriate to append the memory space to the `ArrayView` class rather than the functions in this context. Currently, `ArrayView` is only a wrapper around a pointer plus a size with some access checks, so the approach chosen by @masterleinad makes sense: The pointers do not carry any meaning about the context they are used in, it is the caller who needs to distinguish them.

I do like the fact that `ArrayView` is light-weight so I have a slight preference for what @masterleinad suggests. On the other hand, our concept is not really what e.g. `Kokkos::View` is, but I think that we would overengineer things if we tried to make `ArrayView` robust so that it allows the CPU to directly "look" into GPU memory (which means we copy from device to host in the background, which leads to either terribly slow or terribly complicated implementations).